### PR TITLE
Meta esdt decimals

### PR DESCRIPTION
**Pull request by sasurobert**
Meta esdt decimals.

Testing scenario 1:
Register a semi fungible token and change that to meta ESDT by calling "changeSFTToMetaESDT@0F" the ESDT SC. It will change the tokenType to meta esdt with decimals of 16.

Testing scenario 2:
Register a meta ESDT by calling "registerMetaESDT" with token name, ticker and decimals

Testing scenario 3:
Add tokens to a delegation contract. Any contract. with function "addTokens" from the address which is the config change address of delegation manager.

Testing scenario 4:
AsyncCallback with multi transfer. Deploy contracts - new forwarder. 

Testing scenario 5:
AsnycCall with multi transfer and callback with multi transfer through new forwarder.

Testing scenario 6:
Stake a set of nodes as validator. UnStake a few of them - from the middle. Create a new delegation contract from validator data before epoch 6. Call getAllNodeStates and you will see some problems with the node status. You can compare the view with getBlsKeysStatus from validator contract.
After epoch 6 call the new function: "correctNodesStatus" which will rework the list.

Testing scenario 7:
Do the same as scenario 6 - and there will be no more problems with node status

### CODE REVIEW

File: vm/factory/systemSCFactory.go [hunk 1]
```diff
@@ -249,6 +249,11 @@ func (scf *systemSCFactory) createGovernanceContract() (vm.SystemSmartContract,
 }
 
 func (scf *systemSCFactory) createDelegationContract() (vm.SystemSmartContract, error) {
+	addTokensAddress, err := scf.addressPubKeyConverter.Decode(scf.systemSCConfig.DelegationManagerSystemSCConfig.ConfigChangeAddress)
```
Comment (iulianpascalau) [inline vm/factory/systemSCFactory.go:252]: duplicated L278-L281. Could have been another function that computed these bytes and provide them to createDelegationContract and createDelegationManagerContract

File: vm/systemSmartContracts/esdt.go [hunk 2]
```diff
@@ -353,6 +372,88 @@ func (e *esdt) registerSemiFungible(args *vmcommon.ContractCallInput) vmcommon.R
...
+			numOfDecimals,
+		).Error())
+		return vmcommon.UserError
+	}
+	token, returnCode := e.basicOwnershipChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	if !bytes.Equal(token.TokenType, []byte(core.SemiFungibleESDT)) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:441]: so this is one way process.

File: vm/systemSmartContracts/esdt.go [hunk 3]
```diff
@@ -353,6 +372,88 @@ func (e *esdt) registerSemiFungible(args *vmcommon.ContractCallInput) vmcommon.R
...
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	e.eei.Finish(tokenIdentifier)
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) changeSFTToMetaESDT(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:417]: maybe a short integration test?

**Review by iulianpascalau — COMMENTED** [review]
Good implementation! :+1:

File: vm/systemSmartContracts/esdt.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:441]: exactly.

File: vm/systemSmartContracts/esdt.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:417]: added a set of unit tests for this. can do integration as well.

File: vm/factory/systemSCFactory.go (hunk 1 above)
Comment (sasurobert) [reply vm/factory/systemSCFactory.go:252]: would leave as it is.

File: genesis/process/shardGenesisBlockCreator.go [hunk 4]
```diff
@@ -310,21 +311,15 @@ func createProcessorsForShardGenesisBlock(arg ArgsGenesisBlockCreator, enableEpo
...
-		AheadOfTimeGasUsageEnableEpoch: arg.EpochConfig.EnableEpochs.AheadOfTimeGasUsageEnableEpoch,
-		ArwenV3EnableEpoch:             arg.EpochConfig.EnableEpochs.RepairCallbackEnableEpoch,
-		ArwenChangeLocker:              genesisArwenLocker,
-		ESDTTransferParser:             esdtTransferParser,
+		Config:             arg.VirtualMachineConfig,
+		BlockGasLimit:      math.MaxUint64,
+		GasSchedule:        arg.GasSchedule,
+		ArgBlockChainHook:  argsHook,
+		EpochNotifier:      epochNotifier,
+		EpochConfig:        arg.EpochConfig.EnableEpochs,
```
Comment (iulianpascalau) [inline genesis/process/shardGenesisBlockCreator.go:319]: :+1:

File: vm/systemSmartContracts/esdt_test.go [hunk 5]
```diff
@@ -4156,3 +4156,47 @@ func TestEsdt_ExecuteChangeSFTToMetaESDT(t *testing.T) {
...
+	output = e.Execute(vmInput)
+	assert.Equal(t, vmcommon.Ok, output)
+
+	token, _ = e.getExistingToken(fullTicker)
+	assert.Equal(t, token.NumDecimals, uint32(10))
+	assert.Equal(t, token.TokenType, []byte(metaESDT))
+
+	output = e.Execute(vmInput)
+	assert.Equal(t, vmcommon.UserError, output)
+	assert.True(t, strings.Contains(eei.returnMessage, "change can happen to semi fungible tokens only"))
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt_test.go:4201]: :+1:

File: genesis/process/shardGenesisBlockCreator.go [hunk 6]
```diff
@@ -310,21 +311,15 @@ func createProcessorsForShardGenesisBlock(arg ArgsGenesisBlockCreator, enableEpo
...
+		ArwenChangeLocker:  genesisArwenLocker,
+		ESDTTransferParser: esdtTransferParser,
 	}
-	log.Debug("shardGenesisCreator: enable epoch for sc deploy", "epoch", argsNewVMFactory.DeployEnableEpoch)
```
Comment (iulianpascalau) [inline genesis/process/shardGenesisBlockCreator.go:322]: you removed these epoch activation prints. Can we have them in the NewVMContainerFactory function?

Comment (sasurobert) [reply genesis/process/shardGenesisBlockCreator.go:322]: I removed from shardGenesisBlock creator - but we still have the one from nodeRunner and from node factory. Here, at genesis is not needed.

Comment (sasurobert) [reply genesis/process/shardGenesisBlockCreator.go:322]: and these prints are not more relevant - as they were used in arwen 1.2 and 1.3.

File: vm/systemSmartContracts/delegation.go [hunk 7]
```diff
@@ -2645,6 +2658,19 @@ func (d *delegation) getMetaData(args *vmcommon.ContractCallInput) vmcommon.Retu
 	return vmcommon.Ok
 }
 
+func (d *delegation) addTokens(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/delegation.go:2661]: This method does not add anything, it just returns UserError or Ok

File: vm/systemSmartContracts/esdt.go [hunk 8]
```diff
@@ -228,7 +237,7 @@ func (e *esdt) checkBasicCreateArguments(args *vmcommon.ContractCallInput) vmcom
 		e.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
-	if len(args.Arguments) < 2 {
+	if len(args.Arguments) < 1 {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt.go:240]: This change in checkBasicCreateArguments method from 2 to 1 does not create some backwards compatibility issues

File: vm/systemSmartContracts/esdt.go [hunk 9]
```diff
@@ -979,6 +1080,10 @@ func (e *esdt) basicOwnershipChecks(args *vmcommon.ContractCallInput) (*ESDTData
 		e.eei.AddReturnMessage("not enough gas")
 		return nil, vmcommon.OutOfGas
 	}
+	if len(args.Arguments) < 1 {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt.go:1083]: This change in basicOwnershipChecks method from 2 to 1 does not create some backwards compatibility issues

File: vm/systemSmartContracts/delegation.go (hunk 7 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2661]: yes. but it actually does - accepts eGLD.

File: vm/systemSmartContracts/esdt.go (hunk 8 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:240]: no. it does not create problems.

File: vm/systemSmartContracts/esdt.go (hunk 9 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:1083]: no backward compatibility issues here.

**Review by SebastianMarian — DISMISSED** [review]

File: vm/systemSmartContracts/delegation.go [hunk 10]
```diff
@@ -481,7 +483,11 @@ func (d *delegation) updateDelegationStatusFromValidatorData(
 		case active:
 			dStatus.StakedKeys = append(dStatus.StakedKeys, nodesData)
 		case unStaked:
-			dStatus.UnStakedKeys = append(dStatus.StakedKeys, nodesData)
+			if d.flagAddTokens.IsSet() {
+				dStatus.UnStakedKeys = append(dStatus.UnStakedKeys, nodesData)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:487]: :man_facepalming:

**Review by iulianpascalau — DISMISSED** [review]

File: vm/systemSmartContracts/delegation.go [hunk 11]
```diff
@@ -2321,7 +2321,13 @@ func (d *delegation) correctNodesStatus(args *vmcommon.ContractCallInput) vmcomm
 	unStakedKeys := make([]*NodesData, 0)
 	notStakedKeys := make([]*NodesData, 0)
 
-	allNodesList := createMergedListWithoutDuplicates(status)
+	validatorData, err := d.getValidatorData(args.RecipientAddr)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	allNodesList := createMergedListWithoutDuplicates(status, validatorData.BlsPubKeys)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:2329]: is this backwards compatible?

Comment (iulianpascalau) [reply vm/systemSmartContracts/delegation.go:2329]: great

**Review by iulianpascalau — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**